silkworm/db/snapshots/index/btree_index.cpp

```diff
@@ -0,0 +1,151 @@
+/*
+   Copyright 2024 The Silkworm Authors
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+#include "btree_index.hpp"
+
+#include <utility>
+
+#include <gsl/util>
+
+#include <silkworm/core/common/endian.hpp>
+#include <silkworm/core/common/util.hpp>
+#include <silkworm/infra/common/ensure.hpp>
+
+namespace silkworm::snapshots::index {
+
+BTreeIndex::BTreeIndex(seg::Decompressor& kv_decompressor,
+                       std::filesystem::path index_file_path,
+                       std::optional<MemoryMappedRegion> index_region,
+                       uint64_t btree_fanout)
+    : file_path_(std::move(index_file_path)) {
+    ensure(kv_decompressor.is_open(), "BTreeIndex: KV file decompressor must be opened");
+
+    // Gracefully handle the case of empty index file before memory mapping to avoid error
+    if (std::filesystem::file_size(file_path_) == 0) {
+        return;
```

Comment:
We should unify the construction behavior with Index and RecSplit. For RecSplit it throws in constructor for this case, but Index wrapper doesn't throw in constructor, because RecSplitIndex is a unique_ptr there. Index::reopen_index() tries to create a RecSplit object and throws.

I suggest to refactor this (perhaps in a follow-up) to throw, and add the BTreeIndex feature to Index (or do it via some other approach).

I'm not a fan of partial construction approach, because it makes other methods brittle (requires to do checks in all other methods to avoid crashing on btree_ or data_offsets_ be null).

Reply:
I'm not a fan of partial construction either, it was meant to be refactored but I forgot, so thank you very much for pointing this out. I've now added throwing on empty index